plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,402 @@
+	"github.com/tdrn-org/go-tr064/services/tr64desc/deviceinfo"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/hosts"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wancommonifconfig"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wandslifconfig"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wanpppconn"
+	"github.com/tdrn-org/go-tr064/services/tr64desc/wlanconfig"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const pluginName = "fritzbox"
+
+const defaultTimeout = config.Duration(10 * time.Second)
+
+type Fritzbox struct {
+	URLs    []string        `toml:"urls"`
+	Collect []string        `toml:"collect"`
+	Timeout config.Duration `toml:"timeout"`
+	tls.ClientConfig
+	Log             telegraf.Logger `toml:"-"`
+	deviceClients   []*tr064.Client
+	serviceHandlers map[string]serviceHandlerFunc
+}
+
+func defaultFritzbox() *Fritzbox {
+	return &Fritzbox{
+		URLs:            make([]string, 0),
+		Collect:         []string{"device", "wan", "ppp", "dsl", "wlan"},
+		Timeout:         defaultTimeout,
+		deviceClients:   make([]*tr064.Client, 0),
+		serviceHandlers: make(map[string]serviceHandlerFunc),
+	}
+}
+
+func (*Fritzbox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *Fritzbox) Init() error {
+	if plugin.Log == nil {
+		plugin.Log = logger.New("inputs", pluginName, "")
+	}
+	err := plugin.initDeviceClients()
+	if err != nil {
+		return err
+	}
+	plugin.initServiceHandlers()
+	return nil
+}
+
+func (plugin *Fritzbox) initDeviceClients() error {
+	for _, rawUrl := range plugin.URLs {
+		parsedUrl, err := url.Parse(rawUrl)
+		if err != nil {
+			return err
+		}
+		client := tr064.NewClient(parsedUrl)
+		client.Debug = plugin.Log.Level().Includes(telegraf.Debug)
+		client.Timeout = time.Duration(plugin.Timeout)
+		tlsConfig, err := plugin.TLSConfig()
+		if err != nil {
+			return err
+		}
+		client.TlsConfig = tlsConfig
+		plugin.deviceClients = append(plugin.deviceClients, client)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) initServiceHandlers() {
+	if choice.Contains("device", plugin.Collect) {
+		plugin.serviceHandlers[deviceinfo.ServiceShortType] = plugin.gatherDeviceInfo
+	}
+	if choice.Contains("wan", plugin.Collect) {
+		plugin.serviceHandlers[wancommonifconfig.ServiceShortType] = plugin.gatherWanInfo
+	}
+	if choice.Contains("ppp", plugin.Collect) {
+		plugin.serviceHandlers[wanpppconn.ServiceShortType] = plugin.gatherPppInfo
+	}
+	if choice.Contains("dsl", plugin.Collect) {
+		plugin.serviceHandlers[wandslifconfig.ServiceShortType] = plugin.gatherDslInfo
+	}
+	if choice.Contains("wlan", plugin.Collect) {
+		plugin.serviceHandlers[wlanconfig.ServiceShortType] = plugin.gatherWlanInfo
+	}
+	if choice.Contains("hosts", plugin.Collect) {
+		plugin.serviceHandlers[hosts.ServiceShortType] = plugin.gatherHostsInfo
+	}
+}
+
+func (plugin *Fritzbox) Gather(acc telegraf.Accumulator) error {
+	for _, deviceClient := range plugin.deviceClients {
+		plugin.gatherDevice(acc, deviceClient)
+	}
```

Comment:
Would it make sense to gather each device in its own goroutine?

Reply:
Good point. Implemented via 6043a4f2fe2d8815dd087ca4e13b0ba2ea47859f